### ISSUE TITLE
exportGqlSchemaPath doesn't work and pluginHook type error.

### ISSUE THREAD

**Issue by matgott** [graphile/postgraphile#1593]
### Summary

I'm using postgraphile with apollo server and typescript. The exportGqlSchemaPath property doesn't work because when I run the code with ts-node the schema is not beign exported and when I try to run "tsc" I get a type error from "psotgraphile" package:
```
node_modules/postgraphile/build/postgraphile/pluginHook.d.ts:76:36 - error TS2344: Type 'Extra' 
does not satisfy the constraint 'Record<string, unknown>'.                                      l
  Index signature is missing in type 'Extra'.

76         context: graphqlWs.Context<GraphQLWSContextExtra>;
```

### Steps to reproduce

tsconfig.json:
```
{
  "compilerOptions": {
    "module": "commonjs",
    "declaration": true,
    "removeComments": true,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "allowSyntheticDefaultImports": true,
    "target": "es2017",
    "sourceMap": false,
    "outDir": "./dist",
    "baseUrl": "./",
    "incremental": true
  },
  "include": ["src/**/*"],
  "exclude": ["node_modules", "**/*.spec.ts"]
}
```

index.ts
```
import * as pg from 'pg';
import { makeSchemaAndPlugin } from 'postgraphile-apollo-server';
import { ApolloServer } from 'apollo-server-fastify';
import { ApolloServerPluginDrainHttpServer } from 'apollo-server-core';
import { ApolloServerPlugin } from 'apollo-server-plugin-base';
import fastify, { FastifyInstance } from 'fastify';
import { PostGraphileOptions } from 'postgraphile';

const getPgPool = () =>
  new pg.Pool({
    connectionString: 'postgres://northic:northic@localhost/turnapp',
  });

const fastifyAppClosePlugin = (app: FastifyInstance): ApolloServerPlugin => {
  return {
    async serverWillStart() {
      return {
        async drainServer() {
          await app.close();
        },
      };
    },
  };
};

const startApolloServer = async () => {
  const app = fastify();

  const pgOptions: PostGraphileOptions = {
    dynamicJson: true,
    exportGqlSchemaPath: './src/generated/schema.graphql',
  };

  const { schema, plugin: pgPlugin } = await makeSchemaAndPlugin(
    getPgPool(),
    'public', // PostgreSQL schema to use
    pgOptions,
  );

  const server = new ApolloServer({
    schema,
    plugins: [
      fastifyAppClosePlugin(app),
      ApolloServerPluginDrainHttpServer({ httpServer: app.server }),
      pgPlugin,
    ],
  });

  await server.start();
  app.register(server.createHandler());
  await app.listen(4000);
  console.log(`🚀 Server ready at http://localhost:4000${server.graphqlPath}`);
};

startApolloServer();
```

### Expected results

schema file generated
typescript compiles code

### Actual results

```
node_modules/postgraphile/build/postgraphile/pluginHook.d.ts:76:36 - error TS2344: Type 'Extra' 
does not satisfy the constraint 'Record<string, unknown>'.
  Index signature is missing in type 'Extra'.

76         context: graphqlWs.Context<GraphQLWSContextExtra>;
```

### Additional Notes
postgraphile version: 4.12.8
typescript version: ^4.5.5
node version: 14.17.0

**Comment by benjie** [comment]
The TypeScript issue is fixed in 4.12.9, it was a breaking type change in a semver-minor release made by one of our dependencies.

The `exportGqlSchemaPath` option is only valid in [library usage](https://www.graphile.org/postgraphile/usage-library/), with the Apollo plugin you're actually using [schema usage](https://www.graphile.org/postgraphile/usage-schema/) which supports fewer options. However, you have access to the schema via `const { schema, plugin: pgPlugin } = await makeSchemaAndPlugin(...` so you can add a `writeFile`/`printSchema` command to your code:

```ts
import { printSchema } from "graphql";
import { writeFile } from "fs/promises";

// ...
  const { schema, plugin: pgPlugin } = await makeSchemaAndPlugin(
    getPgPool(),
    'public', // PostgreSQL schema to use
    pgOptions,
  );

  // Add this line:
  await writeFile("path/to/schema.graphql", printSchema(schema) + "\n");
```

**Comment by matgott** [comment]
> The TypeScript issue is fixed in 4.12.9, it was a breaking type change in a semver-minor release made by one of our dependencies.
> 
> The `exportGqlSchemaPath` option is only valid in [library usage](https://www.graphile.org/postgraphile/usage-library/), with the Apollo plugin you're actually using [schema usage](https://www.graphile.org/postgraphile/usage-schema/) which supports fewer options. However, you have access to the schema via `const { schema, plugin: pgPlugin } = await makeSchemaAndPlugin(...` so you can add a `writeFile`/`printSchema` command to your code:
> 
> ```ts
> import { printSchema } from "graphql";
> import { writeFile } from "fs/promises";
> 
> // ...
>   const { schema, plugin: pgPlugin } = await makeSchemaAndPlugin(
>     getPgPool(),
>     'public', // PostgreSQL schema to use
>     pgOptions,
>   );
> 
>   // Add this line:
>   await writeFile("path/to/schema.graphql", printSchema(schema) + "\n");
> ```

Thanks for your reply. I'll update to 4.12.9.

But regarding the exportation, postgraphile-apollo-server uses "createPostGraphileSchema". Also I tried to generate the schema without using apollos plugin and didn't work.

Anyways, now I'm using the CLI for the schema (better for my solution) but maybe you want to check it.

thanks for you help!

**Comment by benjie** [comment]
Yes createPostGraphileSchema is “schema only” usage, it doesn’t support the option you specified - hopefully code like the few lines I wrote out for you above should work (if not, please share the error you get).
